### PR TITLE
HAWQ-1555. Add access interfaces for protocol and format in pluggable storage framework

### PR DESCRIPTION
**Pull request by chiyang10000**
Please help to review the code.
It will not affect the original HAWQ internal process, just serving as a invoker wrappers for the following work in the pluggable storage framework.

### CODE REVIEW

**Comment by interma** [comment]
Hi chiyang
Can you provide some unit/feature tests with you PR?
You can refer the existed tests in here: 
https://github.com/apache/incubator-hawq/tree/master/src/test/feature

Thanks!

File: src/include/access/plugstorage.h [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+ *    table. It needs to visit the format options to get the table type.
+ */
+void getExternalTableTypeList(const char formatType,
+                              List *formatOptions,
+                              int *formatterType,
+                              char **formatterName);
+
+void getExternalTableTypeStr(const char formatType,
+                             char *formatOptions,
+							 int *formatterType,
```
Comment (jiny2) [inline src/include/access/plugstorage.h]: please double check this format

File: src/backend/access/external/Makefile [hunk 2]
```diff
@@ -26,7 +26,7 @@ subdir = src/backend/access/external
 top_builddir = ../../../..
 include $(top_builddir)/src/Makefile.global
 
-OBJS = fileam.o url.o libchurl.o hd_work_mgr.o pxfuriparser.o pxfheaders.o \
+OBJS = fileam.o plugstorage.o url.o libchurl.o hd_work_mgr.o pxfuriparser.o pxfheaders.o \
```
Comment (jiny2) [inline src/backend/access/external/Makefile]: I think it is better to add this new object to the tail of OBJS list

**Comment by jiny2** [comment]
+1 

I think it makes sense to have partial facility code checked in for simplifying the following code merge and delivery work. Therefore, per my understanding, it is not necessary to provide feature test cases in current phase, as all existing HAWQ test cases target end-to-end process. We can do it later when some functions work. Of course test case is a must do.

@chiyang10000 for this delivery please make sure current existing HAWQ feature test cases pass without unexpected errors raised.

File: src/backend/access/external/plugstorage.c [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+	}
+	else
+	{
+		*formatterType = ExternalTableType_Invalid;
+		*formatterName = NULL;
+		elog(ERROR, "undefined external table format type: %c", formatType);
+	}
+}
+
+void getExternalTableTypeStr(const char formatType,
```
Comment (jiny2) [inline src/backend/access/external/plugstorage.c]: We need some descriptions like the function above, please add more comments before delivery

File: src/backend/access/external/plugstorage.c [hunk 4]
```diff
@@ -0,0 +1,533 @@
...
+#include "nodes/value.h"
+#include "parser/parse_func.h"
+
+/*
+ * getExternalTableTypeList
+ *
+ *    Return the table type for a given external table
+ *
+ *    Currently it is an implementation with performance cost due to the
+ *    fact that the format types for TEXT, CSV, and ORC external tables
```
Comment (stanlyxiang) [inline src/backend/access/external/plugstorage.c]: currently there is no type "TEXT" "CSV" and "ORC".

**Comment by stanlyxiang** [comment]
+1

**Comment by chiyang10000** [comment]
Thanks for your guys detailed reviews and comments. I have updated my code.

**Comment by huor** [comment]
Great to wrap up the pluggable storage framework work and start merging the work! +1

**Pull request closed without merging**